### PR TITLE
[setuptools] Update to 74.1.*

### PR DESCRIPTION
**Pull request by srittau**
Closes: #12613

### CODE REVIEW

File: stubs/setuptools/setuptools/msvc.pyi [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+class winreg:
+    HKEY_USERS: Final[int | None]
+    HKEY_CURRENT_USER: Final[int | None]
+    HKEY_LOCAL_MACHINE: Final[int | None]
+    HKEY_CLASSES_ROOT: Final[int | None]
+
+environ: dict[str, str]
+
+class PlatformInfo:
+    current_cpu: ClassVar[str]
```
Comment (srittau) [inline stubs/setuptools/setuptools/msvc.pyi:13]: Maybe `Final` would be more appropriate?

File: stubs/setuptools/setuptools/msvc.pyi [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    @overload
+    def microsoft(self, key: LiteralString, x86: bool = False) -> LiteralString: ...
+    @overload
+    def microsoft(self, key: str, x86: bool = False) -> str: ...  # type: ignore[misc]
+    def lookup(self, key: str, name: str) -> str: ...
+
+class SystemInfo:
+    WinDir: ClassVar[str]
+    ProgramFiles: ClassVar[str]
+    ProgramFilesx86: ClassVar[str]
```
Comment (srittau) [inline stubs/setuptools/setuptools/msvc.pyi:59]: Same here.

**Comment by Avasam** [comment]
I'll do a proper review later. But just quickly letting you know that the msvc file was restored (with a few methods still pruned).

So you can basically restore https://github.com/python/typeshed/blob/bump-setuptools-to-74.0/stubs%2Fsetuptools%2Fsetuptools%2Fmsvc.pyi , and remove the few extra methods stubtest should catch.

**Comment by srittau** [comment]
> I'll do a proper review later. But just quickly letting you know that the msvc file was restored (with a few methods still pruned) in [pypa/setuptools#4626](https://github.com/pypa/setuptools/pull/4626) .
> 
> So you can basically restore https://github.com/python/typeshed/blob/bump-setuptools-to-74.0/stubs%2Fsetuptools%2Fsetuptools%2Fmsvc.pyi , and remove the few extra methods stubtest should catch.

I've stolen the sys.platform check from the old version, but that had a lot of `Incomplete`s that have been fixed in the new version.

File: stubs/setuptools/setuptools/msvc.pyi (hunk 1 above)
Comment (Avasam) [reply stubs/setuptools/setuptools/msvc.pyi:20]: Agreed, `PlatformInfo` looks like some readonly infromation class based on the architecture.

Comment (Avasam) [reply stubs/setuptools/setuptools/msvc.pyi:20]: Agreed, `PlatformInfo` looks like some readonly infromation class based on the architecture.
```suggestion
    curren_cpu: Final[str]
```

File: stubs/setuptools/setuptools/msvc.pyi [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    def windows_sdk(self) -> LiteralString: ...
+    @property
+    def netfx_sdk(self) -> LiteralString: ...
+    @property
+    def windows_kits_roots(self) -> LiteralString: ...
+    @overload
+    def microsoft(self, key: LiteralString, x86: bool = False) -> LiteralString: ...
+    @overload
+    def microsoft(self, key: str, x86: bool = False) -> str: ...  # type: ignore[misc]
+    def lookup(self, key: str, name: str) -> str: ...
```
Comment (Avasam) [inline stubs/setuptools/setuptools/msvc.pyi:64]: These two can take a `StrPath` as the `key` param (and do represent relative paths, registry key paths to be exact).
But I'm pretty sure it's only meant to be a single final path segment, so relative `Path` support seem quite incidental. I think `str` is fine.

File: stubs/setuptools/setuptools/msvc.pyi [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        HKEY_CURRENT_USER: Final[int]
+        HKEY_LOCAL_MACHINE: Final[int]
+        HKEY_CLASSES_ROOT: Final[int]
+    else:
+        HKEY_USERS: Final[None]
+        HKEY_CURRENT_USER: Final[None]
+        HKEY_LOCAL_MACHINE: Final[None]
+        HKEY_CLASSES_ROOT: Final[None]
+
+environ: dict[str, str]
```
Comment (Avasam) [inline stubs/setuptools/setuptools/msvc.pyi:17]: If we expose this internal implementation mock, I'd prefer re-exporting the real type on windows:
```suggestion
if sys.platform == "win32":
    import winreg as winreg 
    from os import environ as environ
else:
    class winreg:
        HKEY_USERS: Final[None]
        HKEY_CURRENT_USER: Final[None]
        HKEY_LOCAL_MACHINE: Final[None]
        HKEY_CLASSES_ROOT: Final[None]

   environ: dict[str, str]
```
It's also still present in https://github.com/python/typeshed/blob/main/stubs/setuptools/%40tests/stubtest_allowlist_darwin.txt , that entry or these mocks should be removed

File: stubs/setuptools/setuptools/msvc.pyi (hunk 2 above)
Comment (Avasam) [reply stubs/setuptools/setuptools/msvc.pyi:69]: Agreed here as well

```suggestion
    WinDir: Final[str]
    ProgramFiles: Final[str]
    ProgramFilesx86: Final[str]
```

File: stubs/setuptools/@tests/stubtest_allowlist.txt [hunk 5]
```diff
@@ -6,6 +6,13 @@ pkg_resources.ZipProvider.loader
 # @classmethod alias not handled correctly by stubtest
 pkg_resources.ZipManifests.load
 
+# More precise type in typeshed (TypedDict vs os._Environ)
+(setuptools.msvc.environ)?
+
+# At runtime, this is a module on Windows, while it's a class on other
+# platforms and in typeshed
+(setuptools.msvc.winreg)?
+
```
Comment (Avasam) [inline stubs/setuptools/@tests/stubtest_allowlist.txt:15]: That's an odd way of writing these. Probably environment-specific failure? I think my comment at the top of `stubs/setuptools/setuptools/msvc.pyi` should address this.

**Review by Avasam — CHANGES_REQUESTED** [review]
Thanks for the more complete stubs whilst adding back this module! A few comments.

**Review by Avasam — APPROVED** [review]
Thanks!

**Pull request merged**